# adaptive_sampling_scripts
Scripts for adaptive sampling analysis

### analyse_RU.py

`analyse_RU.py` is an edited version of the script available [here](https://github.com/SR-Martin/Adaptive-Sequencing-Analysis-Scripts).
This edited script allows removal of multimapping reads from [Minimap2](https://github.com/lh3/minimap2) and filtering based on % identity to the aligning reference.

To use analyse_RU.py, direct it to a read directory containing `fastq_pass` and `fastq_fail` folders containing `fastq.gz` files generated by a basecaller:
```
python analyse_RU.py -f /path/to/read/directory/ -p 0.84 -o /output/prefix -c 1-256 -i /path/to/reference.fasta -t "23F,19A,19F"
```

To specify which sequence(s) the `reference.fasta` to align to, use the `-t` flag using a comman separated list as above. The names must match fasta headers in the `reference.fasta` file.

To specify which channels to separate (e.g. if you have used adaptive sampling on some channels), use `-c`, followed by a hyphen separated pair of integers between 1 and 512.

To change the proportion of an ALIGNMENT (NOTE: not the full read) that must match a reference to be deemed as correct, use the `-p` flag. As default, a multimapping read will assigned to the sequence it has the greatest aligning proportion to (turn this off using `-r`).

By default, both pass and fail reads are used. To use pass only, specify `-b`

This will generate a `_summary.txt` file containing statistics about read mapping and yield, and `_bootstrap.txt`, which contains bootstrap enrichment calculations (change number of bootstrap samples with `-bs`)

### analyse_coverage.py

`analyse_coverage.py` determines the proportion of a reference covered by reads. 

Run with:
```
python analyse_coverage.py -f /path/to/read/directory/ -p 0.84 -o output/prefix -i /path/to/reference.fasta -t "23F,19A,19F"
```

`-p`, `-c`, `-b`, `-r` and `-t` behave the same as with `analyse_RU.py`.

This will output a `X_adaptive_hist.csv` (for channels specified by `-c`) and `X_control_hist.csv` (for all other channels) for each barcode, which details the coverage at each position in the reference. `X` specifies which reference was been aligned to. Additionally, a `_summary.txt` file detailing total bases aligning and the average coverage to each reference sequence is generated.

### split_by_channel.py

`split_by_channel.py` can be used to split reads into two separated files depending on specified channel. For use when adaptive sampling has been used on some but not all channels.

Run with:
```
python split_by_channel.py --infile /path/to/reads.fastq --out /output/prefix --channels 1-256
```

This will output `target.fastq` for reads from the stipulated channels, and `nontarget.fastq` for all others.